### ISSUE TITLE
Create a test configuration framework

### ISSUE THREAD

**Issue by rvs** [lf-edge/eden#10]
Currently, there's a lot of attempts of configuring individual tests via ad-hoc mechanisms like looking up environment variables, etc. A good example is baseOS upgrade test where what baseOS to upgrade to and from can largely be any combination and anything that we would put into the test statically would be a random choice.

We need to come up with a generic framework to feed tests that type of configuration information. Off the top of my head, I could imagine a folder called conf that would be full of json files named according to the testname itself. On a startup each test looks up for the json there and slurps its content to use it as a configuration information. On top of that we can introduce similar schema for the ENV variables to be able to tweak those setting in-place without editing the json file (but that can come in later).

Now, I've just come up with this scheme right now -- so I don't know if there's anything that already exists in the open source that would help us do this.

@deitch your feedback here would be invaluable

**Comment by deitch** [comment]
Can we be a bit clearer on our requirements? Clearly, we are talking about integration tests if we are in Eden. Can we put together some examples of configuration that we want to source? Start with a list here. Even better would be if we could have types of tests those are used for. E.g. "app config", used to test domainmgr, volumemgr, etc.

**Comment by deitch** [comment]
I am trying to get a flow going here that we can have a tool fit into.

**Comment by rvs** [comment]
Well @deitch -- what I have in mind is basically test parameterization. Eve with just a handful of tests we have today, we already have way to much os.Getenv stuff to parameterize them:

```
$ git grep Getenv
baseImage_test.go:      eveBaseRef := os.Getenv("EVE_BASE_REF")
baseImage_test.go:      zArch := os.Getenv("ZARCH")
baseImage_test.go:      HV := os.Getenv("HV")
common.go:      adamIP = os.Getenv("ADAM_IP")
common.go:      adamPort = os.Getenv("ADAM_PORT")
common.go:      adamDir = os.Getenv("ADAM_DIST")
common.go:      adamCA = os.Getenv("ADAM_CA")
common.go:      sshKey = os.Getenv("SSH_KEY")
common.go:      checkLogs = os.Getenv("LOGS") != ""
controller_test.go:             eveCert := os.Getenv("EVE_CERT")
controller_test.go:             serial := os.Getenv("EVE_SERIAL")
```

And I am sure I even missed some "weird" ways of how they get parameterized. What I'm saying is that it perhaps makes sense to externalize all this into a series configuration files of some kind just to keep a handle on how the test can expect its environment to be presented to it.

Another use case here are tests that truly don't care about what versions of the object they work with but they HAVE to know some kind of a version when they run. E.g. a baseOS upgrade test can always try upgrading to an arbitrary version of EVE -- the question is -- *what* version. That'd be a good candidate to put into the test config file set as well.

Stuff like that.

Makes sense?

**Comment by rvs** [comment]
Also @deitch -- tell me if I'm taking this too far, but I feel that this may also benefit from externalizing into a test configuration:

https://github.com/lf-edge/eden/blob/master/tests/integration/common.go#L297

https://github.com/lf-edge/eden/blob/master/tests/integration/common.go#L87

**Comment by deitch** [comment]
> I feel that this may also benefit from externalizing into a test configuration:

Yes and no. On the one hand, it would be nice to get it externalized. On the other, it is specific to one or a series of tests, so might be tied in with that tests or bunch of them.

The bigger issue to me is that those specifics shouldn't go in eden, because then things break in eden when things change in eve (or vice-versa). Rather, they should be part of tests inside eve (even if they require an external framework to run), so that they get changed together.

**Comment by deitch** [comment]
The first point you raised [here](https://github.com/lf-edge/eden/issues/10#issuecomment-612131934) is more important. It is not samples that are hardwired into a test, which might to might not be moved elsewhere, but about configs that are explicitly taken from outside - via CLI, external file that a user is expected to configure, or env var. Those fit squarely within the subject of this issue.

I am not aware of any golang framework that handles this natively. That having been said, I tend to see it handled in one of two ways:

* everything gets whatever config it needs (like we started doing here). I am not in love with it, but it works, and I use it for simpler cases.
* single entry point that gets all config and sets vars in a package or on a struct for a test. I do this for more complex cases. But it really only works if there is some common value. 

In the latter example, I might have an entry point that reads all the env vars, and then sets values on vars in the package, or on a struct, which is used by the tests. This keeps the test packages from needing to know how config was set - all it knows is, "the value I want is in this env var or on this member of a struct" - but it still needs to know to get that. 

The reason I wrote, "it really only works if there is some common value", is that if there are 3 configs for test set A, and 4 for test set B, and 2 for test set C, and never the twain shall meet, then centralizing it didn't buy me all that much. It still helps, since the tests need not think about where to get data from the outside, but it helps less.

Think about how eden itself is written. All of env var parsing or CLI parameters are handled in `cmd/`; from there it mostly just calls things in `pkg/`, which know nothing about config, just about being told, "use value X".
